### PR TITLE
Varvara/feature/cancel subscription

### PR DESCRIPTION
**Pull request by varvarashcherbyna7**

### CODE REVIEW

File: migrations/versions/14a571f23eda_to_subscription_add_checked_date_and_.py
Comment (Taraslysak) [inline migrations/versions/14a571f23eda_to_subscription_add_checked_date_and_.py:1]: Why are there always empty migration files in PRs??? this is very strange

File: naples/main.py [hunk 1]
```diff
@@ -9,10 +13,34 @@
...
+
+
+# @asynccontextmanager
+# async def lifespan(api: FastAPI):
+#     task = asyncio.create_task(update_data())
+#     yield
+#     task.cancel()
+#     await task
+
+# api.router.lifespan_context = lifespan
```
Comment (Taraslysak) [inline naples/main.py]: IMHO, bad idea

File: naples/models/user.py [hunk 2]
```diff
@@ -59,7 +60,16 @@ class User(db.Model, ModelMixin):
...
+    subscriptions: orm.Mapped[list["Subscription"]] = orm.relationship(viewonly=True)
+
+    @property
+    def customer_stripe_id(self):
+        return self.subscriptions[0].customer_stripe_id if self.subscriptions else ""
+
+    @property
+    def subscription(self):
+        # get last saved data in subscriptions
+        return self.subscriptions[-1] if self.subscriptions else None
```
Comment (Taraslysak) [inline naples/models/user.py]: Why do we use different subscriptions here? I suggest we always use the same subscription to get data from . And I suggest it to be the last subscription user has

File: naples/routes/auth.py [hunk 3]
```diff
@@ -18,7 +19,11 @@
 from naples import schemas as s
 from naples.logger import log
 from naples.database import get_db
-from naples.routes.utils import create_stripe_customer
+from naples.routes.utils import (
+    create_stripe_customer,
+    get_product_by_id,
+    save_state_subscription_from_stripe,
```
Comment (Taraslysak) [inline naples/routes/auth.py]: why stripe-related logic is kept in utils and not in stripe controller?

File: naples/routes/auth.py [hunk 4]
```diff
@@ -51,12 +56,25 @@ def login(
...
+        days=CFG.DAYS_BEFORE_UPDATE
+    ):
+        stripe_subscription_data = stripe.Subscription.retrieve(user.subscription.subscription_stripe_id)
+
+        product = get_product_by_id(stripe_subscription_data["items"]["data"][0]["plan"]["id"], db)
+
+        save_state_subscription_from_stripe(user.subscription, product, db)
+
+        log(log.INFO, "Subscription state updated for user [%s]", user.email)
+
```
Comment (Taraslysak) [inline naples/routes/auth.py:76]: Please do not forget to check if this subscription is expired before returning token to user

File: naples/routes/subscription.py [hunk 5]
```diff
@@ -147,18 +151,19 @@ async def webhook_received(
...
-    subscription = event_data["object"]
-
-    db_subscription: m.Subscription | None = get_user_subscription(subscription["customer"], db)
-
-    if not db_subscription:
-        log(log.ERROR, "User subscription not found")
-        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="User subscription not found")
-
     if event_type == "checkout.session.completed":
         session = event_data["object"]
```
Comment (Taraslysak) [inline naples/routes/subscription.py:158]: why do we use plain python objects? al plain python objects should be wrapped in pydantic schemas!

File: naples/routes/subscription.py [hunk 6]
```diff
@@ -147,18 +151,19 @@ async def webhook_received(
...
 
+        # if session["mode"] == "subscription":
+        #     subscription = stripe.Subscription.retrieve(session["subscription"])
+
+        #     product = get_product_by_id(subscription["plan"]["id"], db)
+
+        #     user_subscription = save_state_subscription_from_stripe(subscription, product, db)
+
+        #     log(log.INFO, "User subscription has status completed [%s]", user_subscription)
+
```
Comment (Taraslysak) [inline naples/routes/subscription.py]: Why is this code commented?

File: naples/routes/subscription.py [hunk 7]
```diff
@@ -265,14 +222,18 @@ def modify_subscription(
 ):
     """Modify subscription"""
 
-    user_subscription = db.scalar(sa.select(m.Subscription).where(m.Subscription.user_id == current_user.id))
+    user_subscription = get_user_last_data_subscription(current_user.customer_stripe_id, db)
```
Comment (Taraslysak) [inline naples/routes/subscription.py]: This can be a `@property` of user model

File: naples/routes/utils.py [hunk 8]
```diff
@@ -28,14 +29,59 @@ def create_stripe_customer(current_user: s.User) -> stripe.Customer:
     return create_stripe_user
 
 
-def get_user_subscription(customer_stripe_id: str, db: Session) -> m.Subscription | None:
+def get_user_last_data_subscription(customer_stripe_id: str, db: Session) -> m.Subscription | None:
```
Comment (Taraslysak) [inline naples/routes/utils.py]: This is a part of stripe controller, not util

File: naples/routes/utils.py [hunk 9]
```diff
@@ -28,14 +29,59 @@ def create_stripe_customer(current_user: s.User) -> stripe.Customer:
...
+    product_db = db.scalar(sa.select(m.Product).where(m.Product.stripe_price_id == product_price))
+
+    if not product_db:
+        log(log.ERROR, "Product not found")
+        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Product not found, please contact support")
+
+    return s.Product.model_validate(product_db)
+
+
+def save_state_subscription_from_stripe(
```
Comment (Taraslysak) [inline naples/routes/utils.py]: This is a part of stripe controller, not util

File: migrations/versions/44448a38efea_.py
Comment (Taraslysak) [inline migrations/versions/44448a38efea_.py:1]: Same here

**Review by Taraslysak — COMMENTED** [review]
Please consider some reafctoring

File: naples/models/user.py (hunk 2 above)
Comment (varvarashcherbyna7) [reply naples/models/user.py]: I agree, I'll fix it

File: naples/routes/subscription.py (hunk 6 above)
Comment (varvarashcherbyna7) [reply naples/routes/subscription.py]: I doubt whether he is needed!

File: naples/routes/utils.py (hunk 8 above)
Comment (varvarashcherbyna7) [reply naples/routes/utils.py]: ок

**Pull request merged**